Add VK_EXT_ycbcr_3plane_16bit_lsb_formats extension (#2709)
KhronosGroup/Vulkan-Docs: proposals/VK_EXT_ycbcr_3plane_16bit_lsb_formats.adoc

// Copyright 2022-2025 The Khronos Group Inc.
//
// SPDX-License-Identifier: CC-BY-4.0

= VK_EXT_ycbcr_3plane_16bit_lsb_formats
:toc: left
:docs: https://docs.vulkan.org/spec/latest/
:extensions: {docs}appendices/extensions.html#
:sectnums:

This extension adds support for 10/12bit {YCbCr} formats used by software
decoders like ffmpeg, dav1d and libvpx.

== Problem Statement

Many popular software video decoders - such as ffmpeg/libav, dav1d and libvpx -
use formats for 10 and 12 bit content that are not yet supported in Vulkan, but

> are for example in Linux DRM and Mesa GLES. These formats long established and

Reviewer: ```suggestion
are supported by Linux DRM and Mesa GLES. These formats are long established and
```

Author: Done

optimized for the given task and it would be very hard for decoders to switch
to formats optimized for hardware video decoders such as P010.

Supporting these formats directly in drivers has various advantages for apps
and system compositors, as it allows to avoid unnecessary copies and/or quality
loss. Further more it can also simplify code-sharing with code-paths for
hardware decoders.

Unfortunately it is not possible to use the existing 16 bit formats such as
VK_FORMAT_G16_B16_R16_3PLANE_420_UNORM as - unlike most formats used by hardware
decoders - the alignment is LSB, not MSB. The bit-values thus need to get

> shifted by 6 or 4 bit.

Reviewer: ```suggestion
shifted by 6 or 4 bits.
```

Many popular software video decoders - such as ffmpeg/libav, dav1d and libvpx -
use formats for 10, 12 and 14 bit content that are not yet supported in Vulkan,
but are supported by Linux DRM and Mesa GLES (at least the 10 and 12 bit ones).
These formats are long established and optimized for the given task and it would
be very hard for decoders to switch to formats optimized for hardware video
decoders such as P010.

Supporting these formats directly in drivers has various advantages for apps
and system compositors, as it allows to avoid unnecessary copies and/or quality
loss. Further more it can also simplify code-sharing with code-paths for
hardware decoders.

Unfortunately it is not possible to use the existing 16 bit formats such as
VK_FORMAT_G16_B16_R16_3PLANE_420_UNORM as - unlike most formats used by hardware
decoders - the alignment is LSB, not MSB. The bit-values thus need to get
shifted by 6, 4 or 2 bits.

== Proposal

Add the formats in question. For any drivers already supporting YCbCr formats
via shaders this should be straight forward. Values in the 0.0 - 1.0 range just
need to get multiplied by 65535.0 / 1023.0 (for 10 bit), 65535.0 / 4095.0 (for

> 12 bit) or 65535.0 / 16383.0 (for 14 bit).

Reviewer 2: This is a tricky assertion to make. Right now, we assume that the X bits are garbage and (mostly) ignored by sampling. I say "mostly" because it's kind of okay if they're not since they're the low bits and any garbage in those bits will show up as noise which likely isn't perceptible to the human eye. With theese formats, however, your assertion that they're easy for shaders by doing a bit of multiplication assumes that the top bits are always zero. While this is fairly easy for software to guarantee when outputting such an image, it's very hard for the driver to defend against. We could make it invalid usage or unknown results if you have non-zero bits in the high bits when sampling but we'd need to be very explicit about that in the API.

Author: Indeed. So on the DRM side we solved that by using `z` instead of `x`: https://github.com/torvalds/linux/blob/master/include/uapi/drm/drm_fourcc.h#L403-L405

We could do the same here by turning
```
VK_FORMAT_X6G10_X6B10_X6R10_3PLANE_420_UNORM_3PACK16_EXT
```

into

```
VK_FORMAT_Z6G10_Z6B10_Z6R10_3PLANE_420_UNORM_3PACK16_EXT
```

which would hopefully make this clear. WDYT?

Reviewer 2: We could do that. Since no one will ever be writing to these formats, it's not like we would need hardware to support that. I'd feel a little better about that than saying it's ignored. Again, I think this is where @fluppeteer will have good opinions. As might @spencer-lunarg since he did the format table in the first place.

Author: > Since no one will ever be writing to these formats

For completeness: these formats are also the native/optimal input formats for common software encoders, meaning there is a chance that at some point we'd want native driver support for writing into shared buffers that can directly be fed into those sw-encoders. Not super likely, just to keep in mind.

Author: Did this for now (`X` -> `Z`).

Reviewer 3: Sorry for the delay - I tried to write this a couple of weeks ago and GitHub ate it.

I don't love Z as a choice for the same depth-related reasons as @gfxstrand (and arguably should we decide to use WXYZ in a format at any point), but it may be the least bad option - N for "null" may be an alternative, but that may conflict with the common writing of "nnnn" to mean a series of bits. As you say, this feels like a 16-bit format with a different decode path for scaling, since we're saying the "must be zero" bits are actually significant and just expected to be zero; I'm not 100% clear that our hardware, at least, would be able to support this, since our hardware matrix transform has limited inputs. At least we avoided using Z for depth. (Fortunately a recent KDFS update can express mandatory 0/1 bits, I hope as is needed here.)

For reference it would be nice to include these formats in the 56.1 descriptions. C.f.:

VK_FORMAT_G10X6_B10X6R10X6_2PLANE_420_UNORM_3PACK16 specifies an unsigned normalized multi-planar format that has a 10-bit G component in the top 10 bits of each 16-bit word of plane 0, and a two-component, 32-bit BR plane 1 consisting of a 10-bit B component in the top 10 bits of the word in bytes 0..1, and a 10-bit R component in the top 10 bits of the word in bytes 2..3, with the bottom 6 bits of each word unused. The horizontal and vertical dimensions of the BR plane are halved relative to the image dimensions, and each R and B value is shared with the G components for which ⌊iG×0.5⌋=[iB]=[iR] and ⌊jG×0.5⌋=[jB]=[jR]. The location of each plane when this image is in linear layout can be determined via vkGetImageSubresourceLayout, using VK_IMAGE_ASPECT_PLANE_0_BIT for the G plane, and VK_IMAGE_ASPECT_PLANE_1_BIT for the BR plane. This format only supports images with a width and height that is a multiple of two.

I don't think I see that in this PR? (Unless they're autogenerated these days.)

Author: > I don't think I see that in this PR? (Unless they're autogenerated these days.)

Ouch, I'm sorry - this was indeed missing and arguably a very important part :sweat_smile: 

[Added that now](https://github.com/KhronosGroup/Vulkan-Docs/compare/f5bec6c23f2f6736ea3bf41156a73a5c94ce9103..143fe5fe7f4efc130bea3fc2342d522e706805eb).

> I don't love Z as a choice for the same depth-related reasons as @gfxstrand (and arguably should we decide to use WXYZ in a format at any point), but it may be the least bad option

Thanks! IIUC that means unless somebody else voices resistance we can go forward with that.

> I'm not 100% clear that our hardware, at least, would be able to support this

For the record: even if the formats can only be supported via a shader fallback path it would still be a useful for multimedia apps, freeing them from having to carry custom shaders themselves.
